src/main/scala/viper/gobra/frontend/info/base/Type.scala

```diff
@@ -210,4 +217,59 @@ object Type {
     * vector storing the receiver's type for methods and mpredicates.
     */
   case class AbstractType(messages: (PNode, Vector[Type]) => Messages, typing: Vector[Type] ==> FunctionT) extends PrettyType("abstract")
+
+  trait TypeNode extends Node {
+    /**
+      * Applies the type parameter substitution f to this type
+      */
+    def substitute(f: PartialFunction[PIdnDef, Type]): this.type = {
+      this.transform({
+        case TypeParameterT(id, _, _) if f.isDefinedAt(id) => f(id)
+      })
+    }
+
+    /**
+      * Returns uninstantiated in the type
+      */
+    def uninstantiatedTypeParameters(symb: SymbolTable.WithTypeParameters): Seq[PIdnDef] = {
+      /**
+        * collect all type parameters in the type and filter out type parameters that are not declared by the symb
+        * (we consider type parameters that are not declared by the symbol as instantiated)
+        */
+      this.deepCollect({
+        case t: TypeParameterT => t.id
+      }).intersect(symb.typeParameters.map(_.id))
+    }
+
+    override def info: Parser.Info = Source.Parser.Unsourced
+
+    override def withChildren(children: Seq[Any], pos: Option[(Position, Position)], forceRewrite: Boolean): this.type = {
+      assert(pos.isEmpty, "The pos argument must be set to nil if called on Gobra nodes.")
+
+      if (!forceRewrite && this.children == children) {
+        this
+      } else {
+        create(children)
+      }
+    }
+
+    private def create(children: Seq[Any]): this.type = {
+      import scala.reflect.runtime.{universe => reflection}
+      val mirror = reflection.runtimeMirror(reflection.getClass.getClassLoader)
+      val instanceMirror = mirror.reflect(this)
+      val classSymbol = instanceMirror.symbol
+      val classMirror = mirror.reflectClass(classSymbol)
+      val constructorSymbol = instanceMirror.symbol.primaryConstructor.asMethod
+      val constructorMirror = classMirror.reflectConstructor(constructorSymbol)
+
+      // Call constructor
+      val newNode = constructorMirror(children: _*)
+      newNode.asInstanceOf[this.type]
+    }
+
+    /**
+      * override toString method to prevent that the toString method of the internal Node [[viper.gobra.ast.internal.Node]] is used
+      */
+    override def toString: String = ???
```

Comment:
Why is this necessary?

Reply:
because otherwise types which do not override toString (if someone forgets to implement it) will call toString of viper.gobra.ast.internal.info which will then throw a match error

Reply:
See my previous remark. I think that TypeNode should not extend the internal Node. Whatever you are trying to achieve has to be done without it.